Basically pulling the master merge plus correcting a mistake

GuillermoCasas opened the pull request.

GuillermoCasas committed:
Renaming of python IF to include 2d cases in laplacian calculation

GuillermoCasas committed:
Solving error with chandelier benchmark

GuillermoCasas committed:
Adding new run_marin_rain_benchmark.py

To run the 2d benchmark

GuillermoCasas committed:
Merge remote-tracking branch 'origin/debug-ethier-benchmark-2' into debug-ethier-benchmark-2

GuillermoCasas committed:
2: Auto stash before merge of "debug-ethier-benchmark-2" and "origin/debug-ethier-benchmark-2"

GuillermoCasas committed:
Merge remote-tracking branch 'origin/debug-ethier-benchmark-2' into debug-ethier-benchmark-2

GuillermoCasas committed:
Changes for debugging pouliot 2D

Commit:
Merge branch 'master' into debug-ethier-benchmark-2

GuillermoCasas committed:
Merge branch 'master' into debug-ethier-benchmark-2

farrufat-cimne commented on applications/swimming_DEM_application/custom_utilities/fields/pouliot_flow_field_2D.cpp:
```diff
@@ -61,11 +66,17 @@ double PouliotFlowField2D::U0DT(const int i)
 {
-    return - 25 * mExpY[i];
+    const double y = mY[i];
+    KRATOS_WATCH(i)
```
Not sure if Kratos_watch would slow the code, but I would get rid of them (at least in release compilation)

farrufat-cimne commented on applications/swimming_DEM_application/python_scripts/cellular_flow/ethier_benchmark_algorithm.py:
```diff
@@ -65,12 +65,15 @@ def PerformZeroStepInitializations(self):
             coor[1]=node.Y
             coor[2]=node.Z
-            self.flow_field.Evaluate(0.0 , coor,vel, 0)
+            print('444444444444444444444444444')
```
Leftover prints for debug

GuillermoCasas committed:
removing debug leftovers

pooyan-dadvand commented on applications/swimming_DEM_application/custom_elements/calculate_gradient_Pouliot_2012_edge.cpp:
```diff
@@ -139,6 +139,7 @@ void ComputeGradientPouliot2012Edge<TDim, TNumNodes>::AddPouliot2012RHS(VectorTy
         for (unsigned int i = 0; i < TDim; ++i){
             //F(TDim * node_e + i) += 2.0 * h_edge_inv * le[i] * vel_component_variation_along_edge;
             F(TDim * node_e + i) = 2.0 * h_edge_inv_2 * le[i] * vel_component_variation_along_edge;
+            KRATOS_WATCH(vel_component_variation_along_edge)
```
Seems to be a watch to remove....

pooyan-dadvand commented on applications/swimming_DEM_application/custom_utilities/fields/pouliot_flow_field_2D.cpp:
```diff
@@ -15,16 +15,16 @@ void PouliotFlowField2D::ResizeVectorsForParallelism(const int n_threads)
 void PouliotFlowField2D::UpdateCoordinates(const double time, const array_1d<double, 3>& coor, const int i_thread)
```
Would you please improve the styling and naming for the arguments? Time, Coordinates, ThreadNumber, ...

GuillermoCasas replied:
I use uncapitalized names in general in my application... I can change this but I would prefer to do it in a planned, organized way. Likewise I use the abbreviation coor quite often. This is also a change that I could do everywhere at once but not in this pull request, if you don't mind.

pooyan-dadvand commented on applications/swimming_DEM_application/custom_utilities/fields/pouliot_flow_field_2D.cpp:
```diff
@@ -15,16 +15,16 @@ void PouliotFlowField2D::ResizeVectorsForParallelism(const int n_threads)
     }
 }
 
 void PouliotFlowField2D::UpdateCoordinates(const double time, const vector<double>& coor, const int i_thread)
```
Same...

GuillermoCasas replied:
I use the style mCapitalLetters for member veriables everywhere. Same answer as above.

pooyan-dadvand commented on applications/swimming_DEM_application/test_examples/candelier_spiral.gid/console_output.txt:
```diff
@@ -0,0 +1,2076 @@
+
```
This file is needed to be in repository?

pooyan-dadvand reviewed (changes requested):
Some minor changes in styling is required....

Commit:
Merge branch 'master' into debug-ethier-benchmark-2

Commit:
Indentation error

Commit:
Bug fix dem main script

Commit:
Deleting files

Commit:
Input parameter fixed

GuillermoCasas committed:
Making the structured mesh generator available from python

This required some changes that affected all the geometries. It should all be checked for soundness!

GuillermoCasas committed:
latest changes for the Poutliot benchmark

GuillermoCasas committed:
Merge remote-tracking branch 'origin/debug-ethier-benchmark-2' into debug-ethier-benchmark-2

GuillermoCasas committed:
Merge branch 'master' into debug-ethier-benchmark-2

pooyan-dadvand commented on kratos/geometries/nurbs_3d.h:
```diff
@@ -355,12 +355,12 @@ template<class TPointType> class NurbsPatchGeometry3D  //modified by Matthias
     virtual ~NurbsPatchGeometry3D() {}
 
-    GeometryData::KratosGeometryFamily GetGeometryFamily()
+    GeometryData::KratosGeometryFamily GetGeometryFamily() const
```
would you please add also the override here?

pooyan-dadvand commented on kratos/geometries/nurbs_3d.h:
```diff
@@ -355,12 +355,12 @@ template<class TPointType> class NurbsPatchGeometry3D  //modified by Matthias
     }
 
-    GeometryData::KratosGeometryType GetGeometryType()
+    GeometryData::KratosGeometryType GetGeometryType() const
```
and other missing overrides like here?

pooyan-dadvand commented on kratos/processes/structured_mesh_generator_process.cpp:
```diff
@@ -49,38 +51,31 @@ namespace Kratos
-		//Element const& r_element = KratosComponents<Element>::Get(mElementName);
+        Check();
 
 		// I cannot do this test because the GetGeometryType is not a constant method! Pooyan.
```
Now you can remove this comment as you are doing the check

pooyan-dadvand commented on kratos/processes/structured_mesh_generator_process.cpp:
```diff
@@ -232,6 +227,71 @@ namespace Kratos
+
+
+    bool StructuredMeshGeneratorProcess::CheckDomainGeometryConnectivityForHexahedra3D8() {
+        using tetrahedra_connectivity_in_cell_type = std::array<std::size_t, 4>;
```
I understand that you are checking for degenerated hexahedron. Isn't it?

GuillermoCasas replied:
I am constructing the tetrahedra the way you do it with the given geometry and checking for negative volumes. I just guessed (and checked in some cases) that this is an appropriate way to check this condition. We can think of a better way, nonetheless.

pooyan-dadvand commented on kratos/processes/structured_mesh_generator_process.cpp:
```diff
@@ -232,6 +227,71 @@ namespace Kratos
+    }
+
+    bool StructuredMeshGeneratorProcess::CheckDomainGeometryConnectivityForQuadrilateral2D4() {
+        return true; // TO DO: write check
```
Would you please implement also this case?

GuillermoCasas replied:
I have now done so

pooyan-dadvand commented on kratos/python/add_processes_to_python.cpp:
```diff
@@ -148,6 +149,10 @@ void  AddProcessesToPython()
     ;
 
+    class_<StructuredMeshGeneratorProcess, bases<Process>, boost::noncopyable >("StructuredMeshGeneratorProcess",
+            init<Geometry< Node<3> >::Pointer, ModelPart&, Parameters&>())
```
This will lead to a memory error!! Because inside the process you are storing a reference not the `Geometry::Pointer`  and if you remove the input geometry inside python before running the process `Execute ` method, it will crash!! you need to pass the const reference.

pooyan-dadvand replied:
To pass the constant you should say the lifetime of the constant with an additional argument: `with_custodian_and_ward<1,2>` which means that hold the argument 1 until this object is alive!

GuillermoCasas replied:
Great! I have tested it by doing del to the geometry before Execure(). Indeed, without the suggested fix you run into a segmentation fault. With the fix it works without problem.

pooyan-dadvand commented on kratos/processes/structured_mesh_generator_process.h:
```diff
@@ -58,10 +58,14 @@ namespace Kratos
 
-	  /// The object is not copyable.
+      StructuredMeshGeneratorProcess(GeometryType::Pointer pGeometry, ModelPart& rOutputModelPart, Parameters& TheParameters):
+          StructuredMeshGeneratorProcess(*pGeometry, rOutputModelPart, TheParameters){KRATOS_WATCH(mNumberOfDivisions)};
```
This is dangerous!! you are keeping a reference to a geometry which can be removed by smart pointer....

pooyan-dadvand reviewed (changes requested):
The python interface would make your code crash!

GuillermoCasas committed:
aesth.

GuillermoCasas committed:
Starting to adapt to jsons some benchmarks and custom parameters

GuillermoCasas committed:
Starting to implement Pooyans requests

pooyan-dadvand reviewed (approved):
:+1:

GuillermoCasas committed:
changes necessary to run make_mesh_ethier benchmark test test

GuillermoCasas committed:
Adding check for the 2d case

The formula is completely analogous to the tetrahedra one. Therefore, it could be all abstracted with a template<GeometryType>k, bu I do not know if it is worth it, plus it might obscure the algorithm unnecesserely.
